### DISCUSSION TITLE
[company-capf] Use its own category to override the default `completion-styles`

### DISCUSSION THREAD

**Discussion by condy0919** [company-mode/company-mode#1212] [Ideas]
If user customizes `completion-styles` with

``` emacs-lisp
(setq completion-styles '(basic partial-completion substring flex))
```

Due to the `flex` completion style

``` emacs-lisp
(defun foo ()
  (:lsp
```

Typing `:lsp` has a noticeable latency if `lsp-mode` is installed.

There is a workaround borrowed from doom-emacs:

``` emacs-lisp
(define-advice company-capf--candidates (:around (func &rest args))
  "Try default completion styles."
  (let ((completion-styles '(basic partial-completion)))
    (apply func args)))
```

Ideally, `company-capf` should have its own category, e.g.

``` emacs-lisp
(add-to-list 'completion-category-overrides
             ;; or `company-capf'?
             '(company (styles basic partial-completion)))
```

to override the defaults.

**Comment by dgutov** [comment]
If `company-capf` has its own "category" (not really a category, but effectively overriding the default for all completions going through it), when would your customized value of `completion-styles` be used?

**Reply by condy0919** [reply]
> If `company-capf` has its own "category" (not really a category, but effectively overriding the default for all completions going through it), when would your customized value of `completion-styles` be used?

When writing elisp code.

The defualt capf in elisp-mode is

```elisp
elisp-completion-at-point
```

![image](https://user-images.githubusercontent.com/4024656/133506525-7ee5431e-2aec-44e9-9fd3-066249869606.png)

The left one has tons of completions due to `flex` completion style, but it's not intended.

**Reply by dgutov** [reply]
OK then. So you want Elisp completion to use a specific category, so to speak.

Then you can do something like:

```lisp
(add-hook 'emacs-lisp-mode-hook #'my/setup-elisp-category)

(defun my/setup-elisp-category ()
  (setq-local completion-styles '(basic partial-completion)))
```

Or vice versa: if you customized `completion-styles` for use in certain other completions, those completions might actually have defined categories.

**Reply by condy0919** [reply]
> ```lisp
> (add-hook 'emacs-lisp-mode-hook #'my/setup-elisp-category)
>
> (defun my/setup-elisp-category ()
>   (setq-local completion-styles '(basic partial-completion)))
> ```

The hook mechanism works in certain modes.

To be specific, it works in elisp-mode, but failed when using `ielm` which is
derived from `comint-mode`. If the "category" is provided, the hooks can be
avoided.

**Reply by dgutov** [reply]
`ielm` uses `inferior-emacs-lisp-mode`, which runs the corresponding hook.

There is no technical way for `company-capf` to override the category: it is both provided and handled in upstream code. But if your `define-advice` works for your satisfaction, you can of course continue using it.

**Reply by condy0919** [reply]
> `ielm` uses `inferior-emacs-lisp-mode`, which runs the corresponding hook.

Sorry, I have not tested that.

> There is no technical way for `company-capf` to override the category: it is both provided and handled in upstream code. But if your `define-advice` works for your satisfaction, you can of course continue using it.

I mean we can generalize it. If there is no strong objection, I'm happy to raise a PR.

**Reply by dgutov** [reply]
What kind of generalization do you have in mind?

**Comment by vspinu** [comment]
Another use case here. I use `orderless` for my minibuffer completion but I don't want it to interfere with the company completions. Would be great to have a dedicated  `company-capf` category to specifically target it in overrides. 

Or maybe it would make more sense to have `company-completion-styles` global instead? :thinking: 

This is my current workaround:: 

```el
  (setq completion-styles '(orderless basic))
  (setq completion-category-overrides '((file (styles basic partial-completion))))
  ;; https://www.reddit.com/r/emacs/comments/nichkl/comment/gz1jr3s/
  (defun my-company-completion-styles (capf-fn &rest args)
    (let ((completion-styles '(basic partial-completion)))
      (apply capf-fn args)))
  (advice-add 'company-capf :around #'my-company-completion-styles)
```

**Reply by dgutov** [reply]
It seems like you want to have a "code completion" category that the corresponding completion tables would you.

But what if the completion table `company-capf` is working with is tagged with a category already? How do we avoid overriding it?

Do we want to create a wrapper for every completion table (which would return a different category).

The `advice-add` solution you have above is pretty much what I was going to suggest. Alternatively, though, we'd have one or two `company-capf` specific variables, which it would use to override `completion-category-defaults` and/or `completion-styles`.
